### DISCUSSION TITLE
Feature Request: two Channel Counter

### DISCUSSION THREAD

**Discussion by BastiJoe** [arendst/Tasmota#17673] [Ideas]
Hi all,
I saw many applications for counting gas, water or electricity consumptions on old analog meters with reed contacts or optical sensors.
The biggest challenges I saw, was to deal with debouncing, or magnets/reflectors stopping in sensor position and causing a constant HIGH state.
My idea is to implement a two channel counter.
Meaning:
Instead of installing one reed switch, I install two reed switches on two separate GPIOs.
When Gas is flowing the reed contacts switch high one after the other and switch to low, one after the other. This should help avoid counting errors.
Can a "two channel" counter can be implemented?

What do you guys think?

Best regards
Basti
![image](https://user-images.githubusercontent.com/47536246/211587462-10e217ea-e165-498c-9730-a70c65033d1b.png)

**Comment by Jason2866** [comment]
You have to fight probably with two contacts making issues. Which counter will you trust when having different results?

**Reply by BastiJoe** [reply]
Hi Jason, thanks for your reply. I was not thinking about two counters.
I was thinking about one **single** counter, that counts when **both** GPIOs went high and low again.
Some kind of flipflop, lets say.

We might have to consider two cases:
case 1: there is an overlap
case 2: there is no overlap

I will provide some pseudo code probably tomorrow

**Reply by BastiJoe** [reply]
for case 1 (overlap)
````
void loop (){
p1 = digitalRead(12);
p2 = digitalRead(14);

//state 0: p1 low, p2 low
//state 1: p1 high, p2 low
//state 2: p1 high, p2 high
//state 3: p1 low, p2 high
//state 4 = state 0

prev_status = program_status;

switch (prev_status){
	case 0:
		if (p1 == HIGH) {
			program_status = 1;
		}
	case 1:
		if (p2 == HIGH) {
			program_status = 2;
		}
	case 2:
		if (p1 == LOW) {
			program_status = 3;
		}
	case 3:
		if (p2 == LOW) {
			program_status = 0;
			counter ++
		}
}
```
`

**Comment by barbudor** [comment]
I think you are ré-inventing rotary encoder
That may be a trail to follow

**Reply by BastiJoe** [reply]
Oh, you're right xD

**Reply by BastiJoe** [reply]
Hi Barbudor,
any further thoughts about this topic?
I found a similar discussion here:
https://github.com/arendst/Tasmota/discussions/16002

Would be very nice to have a rotary encoder as a counter without limits. That might fit several applications.

Best regards
Sebastian

**Reply by sfromis** [reply]
As already suggested there, to get a rotary up/down counter without min/max, you can use `SetOption43 1` along with `SetOption98 1` to have rule triggers where "position" 1 means increment, and 0 means decrement, which you can use to update a var, or whatever. Rule example:
```
on Rotary1#Pos1=1 do Add1 1 endon
on Rotary1#Pos1=0 do Sub1 1 endon
```
Not hard to do :smile:

**Reply by mxhf** [reply]
All, I do not seem to be able to get this to work. 

With SetOption98 1 nothing seems to change and the rules
"on Rotary1#Pos1==1 do var1 Increment endon on Rotary1#Pos1==0 do var1 Decrement endon"
do not change var1.

I am using:
Tasmota32 8.4.0.3
a ESP32 NodeMCU
the rotary encoder C38S6G5-600B-G24N
pins 32 and 33 
09:50:21 MOD: ESP32-DevKit Module, GPIO0 0, GPIO1 0, GPIO2 0, GPIO3 0, GPIO4 0, GPIO5 0, GPIO12 0, GPIO13 0, GPIO14 0, GPIO15 0, GPIO16 0, GPIO17 0, GPIO18 0, GPIO19 0, GPIO21 0, GPIO22 0, GPIO23 0, GPIO25 0, GPIO26 0, GPIO27 0, GPIO32 3264, GPIO33 3296, GPIO34 0, GPIO35 0, GPIO36 0, GPIO39 0

Encoder hooked up to GND and 5V (shield also connected to GND.
Now, using the COUNTER on pin 32 or 33 works.
I do see a DIMMER change when I do not use SetOption98 1. E.g.
09:48:43 MQT: stat/tasmota_AC231C/RESULT = {"POWER":"OFF","Dimmer":40}
09:48:44 MQT: stat/tasmota_AC231C/RESULT = {"POWER":"OFF","Dimmer":0}

I have programmed a little sketch that also verified that I can read the encoder correctly.

Minor comment. Upon setting pins 32 and 33  to RotaryA/B the GUI is not showing 
any value, is this normal?

Any help?

**Reply by sfromis** [reply]
It is not clear whether you entered *and* activated the rule correctly, and neither if you did *both* SetOption commands. What is the output of this command?
`Backlog Rule0;SetOption43;SetOption98`

Also, you are using a *very* old Tasmota version, I'd recommend upgrading to current, as I don't remember when what enhancements were made.

**Reply by BastiJoe** [reply]
I finally installed this on my own but i have issues setting up the script. 
Unfortunately i cannot use rules because of SML usage.
Could you check my script please? :) 
@barbudor 

Thank you very much in advance.

BR
Sebastian

![image](https://github.com/user-attachments/assets/dadbbdfd-d21c-4cf9-b43a-d2f841a0abe9)

```
>D
>B
=>sensor53 r
=>sensor53 l2
;Set teleperiod to 20sec  
;tper=10  


# Initialize variables
encoderstate = 0
counter=0  # New counter initialized to 0

>M 1
+1,3,s,0,9600,Power
1,77070100600100ff@#,Zählernummer,,Meter_Number,0
1,77070100010800ff@1000,Verbrauch,kWh,Total_in,4
1,77070100100700ff@1,Leistung,W,Power_curr,0
1,77070100020800ff@1000,Erzeugung,kWh,Total_out,4
#




>F
# Check button states
if sw[1]==1  
then  
  encoderstate=1
  print encoder state 1
endif
if encoderstate==1
then
if sw[2]==1  
then  
  encoderstate=2
  print encoder state 2
endif
endif

if encoderstate==2
then
  if sw[2]==0
  then
    # Increment the counter when both buttons are pressed
    counter=counter+1
    print gas counter increase

    # Publish GAS data with the current counter value
    =>Publish tele/%topic%/RESULT {"GAS":1, "Counter":counter}
    
    # Reset button states
    encoderstate = 0
    print resetting encoder state
  endif
endif
```

**Comment by mxhf** [comment]
OK, I upgraded to 12.4.0

I am getting
11:15:30.033 CMD: Backlog Rule0;SetOption43;SetOption98
11:15:30.070 MQT: stat/tasmota_AC231C/RESULT = {"Rule1":{"State":"OFF","Once":"OFF","StopOnError":"OFF","Length":85,"Free":426,"Rules":"on Rotary1#Pos1==1 do var1 Increment endon on Rotary1#Pos1==0 do var1 Decrement endon"}}
11:15:30.089 MQT: stat/tasmota_AC231C/RESULT = {"Rule2":{"State":"OFF","Once":"OFF","StopOnError":"OFF","Length":0,"Free":511,"Rules":""}}
11:15:30.107 MQT: stat/tasmota_AC231C/RESULT = {"Rule3":{"State":"OFF","Once":"OFF","StopOnError":"OFF","Length":0,"Free":511,"Rules":""}}
11:15:30.287 MQT: stat/tasmota_AC231C/RESULT = {"SetOption43":1}
11:15:30.537 MQT: stat/tasmota_AC231C/RESULT = {"SetOption98":"ON"}

**Reply by sfromis** [reply]
You did not activate the rule set. `"State":"OFF",`
`Rule1 1` to do that.

**Reply by mxhf** [reply]
WORKS! (Feeling very stupid!) Could you give me a quick rule to push out the var1 via MQTT as part of the telemetry?
